src/test/clone/test_fork.rs

```diff
@@ -446,6 +978,438 @@ fn main() -> Result<(), Box<dyn Error>> {
+                        assert_eq!(rv, 0);
+                        Pid::from_raw(unsafe { info.si_pid() })
+                    }
+                    Err(nix::errno::Errno::ECHILD) => None,
+                    Err(e) => panic!("Unexpected error {e:?}"),
+                }
+            })
+        },
+        all_envs.clone(),
+    ));
+
+    tests.push(ShadowTest::new(
+        "test_waitfn_honors_wnohang:waitpid",
+        || {
+            test_waitfn_honors_wnohang(|options| {
+                nix::errno::Errno::result(unsafe {
+                    libc::waitpid(-1, std::ptr::null_mut(), options)
+                })
+            })
+        },
+        all_envs.clone(),
+    ));
+    tests.push(ShadowTest::new(
+        "test_waitfn_honors_wnohang:waitid",
+        || {
+            test_waitfn_honors_wnohang(|options| {
+                nix::errno::Errno::result(unsafe {
+                    libc::waitid(
+                        libc::P_ALL,
+                        0,
+                        std::ptr::null_mut(),
+                        options | libc::WEXITED,
+                    )
+                })
+            })
+        },
+        all_envs.clone(),
+    ));
+
+    /* waitpid does *not* implement WNOWAIT */
+    /*
+    tests.push(ShadowTest::new(
+        "test_waitfn_honors_wnowait:waitpid",
+        || {
+            test_waitfn_honors_wnowait(|options| {
+                match nix::errno::Errno::result(unsafe {
+                    libc::waitpid(-1, std::ptr::null_mut(), options)
+                }) {
+                    Ok(pid) => Some(Pid::from_raw(pid).unwrap()),
+                    Err(nix::errno::Errno::ECHILD) => None,
+                    Err(e) => panic!("Unexpected error: {e:?}"),
+                }
+            })
+        },
+        all_envs.clone(),
+    ));
+    */
+    tests.push(ShadowTest::new(
+        "test_waitfn_honors_wnowait:waitid",
+        || {
+            test_waitfn_honors_wnowait(|options| {
+                let mut info: siginfo_t = unsafe { std::mem::zeroed() };
+                match nix::errno::Errno::result(unsafe {
+                    libc::waitid(libc::P_PGID, 0, &mut info, options | libc::WEXITED)
+                }) {
+                    Ok(rv) => {
+                        assert_eq!(rv, 0);
+                        Pid::from_raw(unsafe { info.si_pid() })
+                    }
+                    Err(nix::errno::Errno::ECHILD) => None,
+                    Err(e) => panic!("Unexpected error {e:?}"),
+                }
+            })
+        },
+        all_envs.clone(),
+    ));
+
+    for use_wclone in [true, false] {
+        tests.push(ShadowTest::new(
+            &format!(
+                "test_waitfn_honors_wclone:waitid:{}",
+                if use_wclone { "set" } else { "unset" }
+            ),
+            move || {
+                test_waitfn_honors_wclone(use_wclone, |options| {
+                    let mut info: siginfo_t = unsafe { std::mem::zeroed() };
```

Comment:
Maybe add some comments around the new tests? These are testing some pretty specific linux things that someone looking at the code might not know much about (myself included), so having more comments will help if this breaks in the future.

Reply:
Done